### PR TITLE
ENH: optimize: change default linprog method to interior-point

### LINKED ISSUE #9636: linprog benchmarking and proposal to make 'interior-point' the default method
In support of #9263, I benchmarked the new revised simplex implementation against the other `scipy.optimize.linprog` methods and two other open source LP solvers with Python interfaces: [CVXOPT ](https://cvxopt.org/) and [GLPK](https://www.gnu.org/software/glpk/). I ask the community for feedback on this PR and suggest that in addition to adding the revised simplex method, interior-point should replace the original simplex as the new *default* method for linprog.

### Background
To my knowledge, CVXOPT is the only other popular, open source linear programming solver with a Python interface that is actually *implemented* in Python. GLPK is implemented in ANSI C but CVXOPT provides a Python interface. Both are released under the GNU GPL 3. 

I intended to test [COIN CBC](https://projects.coin-or.org/Cbc), but of the four Python interface options I could find, [PuLP ](https://pythonhosted.org/PuLP/) and [Pyomo ](http://www.pyomo.org/) could not directly interpret input from MPS files or numpy arrays, [CyLP](https://github.com/coin-or/CyLP) is incompatible with Python 3, and installation of [yaposib](https://pypi.org/project/yaposib/), which has not been updated since 2012, failed.

Other popular open source LP solvers exist, but testing was restricted to those with Python interfaces, and of those GLPK and COIN CBC were the fastest and most robust according to [this report](https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2013/138847.pdf) prepared at Sandia National Laboratories in late 2013.

### Methods
All linprog methods, CVXOPT, and GLPK were tested on the 92 Netlib problems without `RANGES` sections in their definitions. (RANGES is just an alternative way of expressing inequality constraints in the MPS format; most of the Netlib problems express inequality constraints otherwise.) Tests were performed on Macbook Pro (Early 2013 , 2.6 GHz intel Core i5, 8 GB 1600 MHz DDR3 RAM) with macOS 10.13.6 and the Anaconda distribution of Python 3.6.6. 

`scipy.optimize.linprog` was tested in a virtual environment with a development version of SciPy and relevant dependencies NumPy 1.15.0, MKL 2018.0.3, and SuiteSparse 5.2.0. This virtual environment was cloned for testing of CVXOPT/GLPK, as installing CVXOPT seemed to break my Spyder installation in the past. Installation of CVXOPT according to the [official instructions](http://cvxopt.org/install/) downgraded SuiteSparse to 4.5.6 and linked NumPy and SuiteSparse to OpenBLAS 0.3.3.

Solution times were measured using `time.perf_counter`; they include the time of loading the problem from the relevant MPS (GLPK and CVXOPT) or .npz file (linprog) and calling the solver within a `try`-`except` block.

### Results

**Dense Solvers**
Although the Netlib benchmark problems tend to be quite sparse (see [problem summary table](http://www.netlib.org/lp/data/readme)), I solve them with dense matrix techniques options to estimate performance on problems that are inherently dense. Results for GLPK are not shown because GLPK does not seem to use dense techniques when controlled via CVXOPT's interface. That is, CVXOPT exposes an option that is supposed to provide this control, but GLPK is consistently faster in 'dense' mode than 'sparse' mode for every problem. This and the fact that the speed difference is usually quite small suggests that only one solution technique is actually being used, and given its speed, it is likely employing sparse matrix techniques.

*Speed*
Bar graphs show elapsed time in seconds for each problem and each solver if:

- the solver reported success, and 
- the reported value of the objective function was within 0.1% of the accepted value.

Otherwise, the value shown is 0.
Each colored bar represents a combination of a problem, indicated by the bar's horizontal location, and solver, indicated by the bar's color; the elapsed time is represented by the position of the _top_ of each bar on the vertical axis. (This is _not_ the visible length of each bar; the bars are layered in order of speed, as if in a third dimension, with the fastest solver - and shortest bar - for the particular problem in front.)

<div>
    <a href="https://plot.ly/~mdhaber/104/?share_key=Qr3EsljqwnMBSGP4xAF0Xv" target="_blank" title="Dense LP Solver Benchmarking" style="display: block; text-align: center;">Dense LP Solver Benchmarking</a>
</div>

 - linprog (interior point) is faster than any other solver for every problem it solves.
 - linprog (simplex) solves only a few problems. For each of these, linprog (revised simplex) is faster.
 - Of the problems solved by both linprog (revised simplex) and CVXOPT, linprog (revised simplex) tends  to be faster for problems that take more than a second to solve.
 - There are a few problems that linprog (revised simplex) solves that no other solver solves.
 - There are a few problems that only CVXOPT solves with default options. Note, however, that at least one SciPy solver can solve each of these with appropriate settings.

 
*Reliability*
Pie charts depict the percentage of problems for which the solver:

- reported success and achieved reasonable accuracy (True Positive),
- reported failure yet achieved reasonable accuracy (False Negative),
- reported failure and failed to achieve reasonable accuracy (True Negative), and
- reported success yet failed to achieve reasonable accuracy (False Positive),

where 'reasonable accuracy' means that the objective function is within 0.1% of the accepted value.

<div>
    <a href="https://plot.ly/~mdhaber/106/?share_key=oNuP6TSKuRiWNn06BxQnp8" target="_blank" title="Dense LP Solver Benchmarking Reliability" style="display: block; text-align: center;">Dense LP Solver Benchmarking Reliability</a>
</div>

* linprog (interior-point) is the most reliable.
* There is a considerably percentage of problems for which linprog (interior-point) reports failure but is reasonably accurate. This typically occurs because linprog (interior-point)'s internal criterion for success is much stricter than the 0.1% tolerance used here.
* linprog (revised simplex) is arguably less reliable than CVXOPT, but certainly more reliable than linprog (simplex)

**Sparse Solvers**
The results for the sparse solvers are represented in the same way.  Due to the apparent error in CVXOPT's interface to GLPK, results are for GLPK with the 'dense' option as it is always faster than the 'sparse' option.

*Speed*
<div>
    <a href="https://plot.ly/~mdhaber/108/?share_key=cJRgt4ZsFXDOhuqy0sxjaZ" target="_blank" title="Sparse LP Solver Benchmarking" style="display: block; text-align: center;">Sparse LP Solver Benchmarking</a>
</div>

- Of the problems solved by both linprog (interior point) and CVXOPT, linprog (interior point) is typically faster except for the smallest problems.
- Of the problems solved by both linprog (interior point) and GLPK, linprog(interior-point) is often faster, but not always.

*Reliability*

<div>
    <a href="https://plot.ly/~mdhaber/110/?share_key=ymKcQe0UqH6WKxBOXVd2AG" target="_blank" title="Sparse LP Solver Benchmarking Reliability" style="display: block; text-align: center;">Sparse LP Solver Benchmarking Reliability</a>
</div>

* GLPK is the most reliable, but linprog (interior-point) is quite reliable, too. Both are considerably more reliable than CVXOPT.
* There is a considerably percentage of problems for which linprog (interior-point) reports failure but is reasonably accurate. This typically occurs because linprog (interior-point)'s internal criterion for success is much stricter than the 0.1% tolerance used here.

### Discussion and Conclusion
Although the methods in `scipy.optimize.linprog` are not always as fast and reliable as GLPK, the sparse interior point option certainly rivals GLPK's speed. Also, a closer look at the failure state messages for the interior point options suggests that linprog's internal success criterion may be too strict, as it frequently reports failure even though it achieves reasonable accuracy. Adjusted appropriately, the sparse solver may approach GLPK in terms of reliability. In any case, the interior point options are significantly faster and more reliable than CVXOPT, the only other popular open source linear programming solver implemented in Python. Although there is much room for improvement, the new dense revised simplex method is a valuable addition to linprog because it 

1. provides *basic* (vertex of the polytope defined by the constraints) solutions to a wider variety of problems faster than the original simplex implementation, 
2. offers greater accuracy than the interior point options, and 
3. solves some problems for which all other methods fail or report failure.

An elusive bug in the linprog docstring is causing #9263 the documentation build test to fail, but I suggest that once it is passing, a final commit should make 'interior point' the default linprog method, and #9263 should be merged.

_This work was supported by a NumFOCUS Small Development Grant_

### PR DESCRIPTION
**Pull request by mdhaber**
Closes #9636

### CODE REVIEW

**Comment by mdhaber** [comment]
I'll fix the doctest issue (add `# may very` as needed) when the Travis CI `FAIL scipy/stats/tests/test_stats.py::TestIQR::test_scale` has been addressed, as I'd prefer to see the build passing _all_ tests before merge.

File: scipy/optimize/_linprog.py [hunk 1]
```diff
@@ -460,12 +460,27 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
...
+       slack: array([3.89999997e+01, 8.46872439e-08] # may vary
+      status: 0
+     success: True
+           x: array([ 9.99999989, -2.99999999]) # may vary
+
+    Note that the default method for `linprog` is 'interior-point', which is
+    approximate by nature. If you need greater accuracy, try 'revised simplex'.
+
+    >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds],
+                      method = 'revised simplex')
```
Comment (Kai-Striega) [inline scipy/optimize/_linprog.py]: This should be ``method='revised simplex'`` *not* ``method = 'revised simplex'`` (note the extra spaces).

File: scipy/optimize/_linprog.py [hunk 2]
```diff
@@ -460,12 +460,27 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
...
+         con: array([], dtype=float64)
+         fun: -21.99999984082494 # may vary
+     message: 'Optimization terminated successfully.'
+         nit: 6 # may vary
+       slack: array([3.89999997e+01, 8.46872439e-08] # may vary
+      status: 0
+     success: True
+           x: array([ 9.99999989, -2.99999999]) # may vary
+
+    Note that the default method for `linprog` is 'interior-point', which is
```
Comment (Kai-Striega) [inline scipy/optimize/_linprog.py]: Maybe add the note before printing the result? I suspect quite a few people will see the "wrong" answer and stop reading there.

File: scipy/optimize/_linprog.py (hunk 1 above)
Comment (mdhaber) [reply scipy/optimize/_linprog.py]: Right.

File: scipy/optimize/_linprog.py (hunk 2 above)
Comment (mdhaber) [reply scipy/optimize/_linprog.py]: OK.

File: scipy/optimize/_linprog.py [hunk 3]
```diff
@@ -460,6 +460,9 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
     >>> x1_bounds = (-3, None)
     >>> from scipy.optimize import linprog
     >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds])
+
+    Note that the default method for `linprog` is 'interior-point', which is
+    approximate by nature.
```
Comment (Kai-Striega) [inline scipy/optimize/_linprog.py]: Here you will need to insert a blank line after line 465.

File: scipy/optimize/_linprog.py [hunk 4]
```diff
@@ -470,11 +473,8 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
      success: True
            x: array([ 9.99999989, -2.99999999]) # may vary
 
-    Note that the default method for `linprog` is 'interior-point', which is
-    approximate by nature. If you need greater accuracy, try 'revised simplex'.
-
-    >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds],
-                      method = 'revised simplex')
+    If you need greater accuracy, try 'revised simplex'.
+    >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds], method='revised simplex')
```
Comment (Kai-Striega) [inline scipy/optimize/_linprog.py]: I think you also need a blank line here.

Comment (mdhaber) [reply scipy/optimize/_linprog.py]: Thanks!

File: scipy/optimize/_linprog.py (hunk 3 above)
Comment (mdhaber) [reply scipy/optimize/_linprog.py]: Added. Hope this fixed it.

**Comment by mdhaber** [comment]
@Kai-Striega Thank you!
I hope that worked but I can't see because there seems to be an unrelated failure. Do you agree that it looks unrelated?
Same in #9701, #9694, maybe more.

**Comment by Kai-Striega** [comment]
@mdhaber the error occurred on travis-ci, which is now passing. I don't think the new error is related, it looks like a build/NumPy issue. It may still be worth waiting until it's fixed just to be sure.

**Comment by mdhaber** [comment]
@kai-striega oops I forgot it was on Travis CI.
Yes, let's wait. Also @rgommers should take a look since it's a bit of a backwards compatibility break.

**Comment by mdhaber** [comment]
I just can't catch a break.

**Comment by mdhaber** [comment]
Much better.

**Comment by Kai-Striega** [comment]
@mdhaber glad it's finally passing. Another minor point, I've been looking at the NumPy doc guide regarding the input documentation. It says: 

> When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first:

    order : {'C', 'F', 'A'}
        Description of `order`.

Whereas the default for linprog is written: 

https://github.com/scipy/scipy/blob/5acfd3fc50fa297245de0161e05aee62d61beb66/scipy/optimize/_linprog.py#L222-L227

I've been looking at tidying the documentation a little in another PR, but if it's not too much trouble would you change it here?

**Comment by mdhaber** [comment]
@Kai-Striega adjusted

**Comment by Kai-Striega** [comment]
@mdhaber looking good,  from my perspective this is ready to merge.

**Comment by mdhaber** [comment]
@rgommers OK to change `linprog` default to `'interior-point'`?

**Comment by rgommers** [comment]
Yep, LGTM

**Comment by Kai-Striega** [comment]
Merged =). Thanks for putting the time into this.

**Pull request merged**